floors/enforce.go

```diff
@@ -0,0 +1,202 @@
+			if !ok {
+				continue
+			}
+
+			reqImpCur := reqImp.BidFloorCur
+			if reqImpCur == "" {
+				reqImpCur = defaultCurrency
+				if bidRequestWrapper.Cur != nil {
+					reqImpCur = bidRequestWrapper.Cur[0]
+				}
+			}
+			updateBidExtWithFloors(reqImp, bid, reqImpCur)
+
+			if !isPriceFloorsEnforcementDisabled(bidRequestWrapper) {
+				dealBid := checkDealBidForEnforcement(bid, enforceDealFloors)
+				if dealBid != nil {
+					eligibleBids = append(eligibleBids, dealBid)
+					continue
+				}
+
+				rate, err := getCurrencyConversionRate(seatBid.Currency, reqImpCur, conversions)
+				if err != nil {
+					errs = append(errs, fmt.Errorf("error in rate conversion from = %s to %s with bidder %s for impression id %s and bid id %s error = %v", seatBid.Currency, reqImpCur, bidderName, bid.Bid.ImpID, bid.Bid.ID, err.Error()))
+					continue
+				}
+
+				bidPrice := rate * bid.Bid.Price
+
+				if reqImp.BidFloor > bidPrice {
+					rejectedBid := &entities.PbsOrtbSeatBid{
+						Currency: seatBid.Currency,
+						Seat:     seatBid.Seat,
+						Bids:     []*entities.PbsOrtbBid{bid},
+					}
+					rejectedBids = append(rejectedBids, rejectedBid)
+					continue
+				}
+			}
+			eligibleBids = append(eligibleBids, bid)
+		}
+		seatBids[bidderName].Bids = eligibleBids
+	}
+	return seatBids, errs, rejectedBids
+}
+
+// isPriceFloorsEnforcementDisabled check for floors are disabled at request using Floors.Enforcement.EnforcePBS flag
+func isPriceFloorsEnforcementDisabled(bidRequestWrapper *openrtb_ext.RequestWrapper) bool {
+	requestExt, err := bidRequestWrapper.GetRequestExt()
+	if err == nil {
+		if prebidExt := requestExt.GetPrebid(); prebidExt != nil && prebidExt.Floors != nil && !prebidExt.Floors.GetEnforcePBS() {
+			return true
+		}
+	}
+	return false
+}
+
+// isFloorsSignallingSkipped check for floors signalling is skipped due to skip rate
+func isFloorsSignallingSkipped(requestExt *openrtb_ext.RequestExt) bool {
+	if requestExt != nil {
+		if prebidExt := requestExt.GetPrebid(); prebidExt != nil && prebidExt.Floors != nil {
+			return prebidExt.Floors.GetFloorsSkippedFlag()
+		}
+	}
+	return false
+}
+
+// getEnforceRateRequest returns enforceRate provided in request
+func getEnforceRateRequest(requestExt *openrtb_ext.RequestExt) int {
+	if requestExt != nil {
+		if prebidExt := requestExt.GetPrebid(); prebidExt != nil && prebidExt.Floors != nil {
+			return prebidExt.Floors.GetEnforceRate()
+		}
+	}
+	return 0
+}
+
+// getEnforceDealsFlag returns FloorDeals flag from req.ext.prebid.floors.enforcement
+func getEnforceDealsFlag(requestExt *openrtb_ext.RequestExt) bool {
+	if requestExt != nil {
+		if prebidExt := requestExt.GetPrebid(); prebidExt != nil && prebidExt.Floors != nil {
+			return prebidExt.Floors.GetEnforceDealsFlag()
+		}
+	}
+	return false
+}
+
+// isValidImpBidFloorPresent checks if non zero imp.bidfloor is present in request
+func isValidImpBidFloorPresent(bidRequest *openrtb2.BidRequest) bool {
```

Comment:
Instead of passing the entire bid request, can we just pass this function what it needs by passing in `[]openrtb2.Imp`?

Reply:
done